src/ssl/support.cc

```diff
@@ -657,7 +586,7 @@ Ssl::InitClientContext(Security::ContextPointer &ctx, Security::PeerOptions &pee
         }
     }
 
-    maybeSetupRsaCallback(ctx);
+    Ssl::MaybeSetupRsaCallback(ctx);
```

Comment:
What make this practice of implicitly shadowing symbol names better than defining explicitly which symbol is being used?

I am aware that explicit full-naming is not strictly necessary. However doing so increases code clarity, self-documentation and is not harmful AFAIK. It just replicates explicitly what the compiler *should* be doing, without relying on that implicit behaviour. By contrast relying on implicit compiler behaviours opens the code to bugs in the toolchain, or IDE environments auto-spelling correction. So "best" seems a bit strong to use describing the minimalist coding style.

Reply:
> What make this practice of implicitly shadowing symbol names better than defining explicitly which symbol is being used?

Your specific question is too loaded to provide a meaningful answer, but I can answer a similar "Why is it usually best to not explicitly mention namespace X inside namespace X?" question with these points:

1. It is not required. With all other factors being equal, less ink is better.
2. It is not useful: The reader already knows the context from the method namespace (or equivalent).
3. It is harmful: The explicit mention of X may confuse the reader into thinking that the surrounding code is not inside X or that there is some naming conflict that should be resolved by explicitly mentioning X.
4. It complicates code refactoring by requiring more adjustments if the code is moved into a different namespace and/or X is renamed.
5. It increases code inconsistencies because only some developers will add (only) some unnecessary namespace prefixes.

> implicitly shadowing symbol names

There should be no shadowing in most cases:

- If both `X::foo()` and `::foo()` exist, then (in most cases) something went wrong at the design stage because such shadowing is (in most cases) too risky/dangerous since the compiler will not be able to identify the cases of the wrong foo() used.
- When both `X::foo()` and `Y::foo()` exist, one does not shadow the other. This is a quite common case because different libraries/scopes (`X` vs `Y`) often use the same names (`foo`).

Reply:
Still open.

Reply:
IME the only solution to collisions when they do occur is explicit X::foo() and ::foo() use in callers. The various compiler and linker combinations do a bunch of slightly different symbol resolving , so the outcome of which gets picked is not as predictable as one might expect from simple local tests.

I think we may have some collisions between objects built before and after the patching. But that should not be too hard to resolve with a clean build. Sooo going with your way for now. I'd hate to be the one debugging it though. FTR; if this does go wrong the potential side effect is OpenSSL not calling our callbacks and doing its own thing to verify a TLS/SSL connections ciphers.

Reply:
Addressed (as far as this particular instance of unnecessary namespace prefixes is concerned; I will try to mark several other occurrences separately).

> the only solution to collisions when they do occur is explicit X::foo() and ::foo() use in callers

IMO, we should not rely on humans to remember about (and detect/fix) silent collisions when writing innocent caller code. We should instead design our APIs so that undetected collisions become nearly impossible. If both X::foo() and ::foo() may be present (and can compile) inside the same reasonable caller code (leading to two different results), then something went wrong at the _design level_. Such a problem should be solved at the design level (instead of adding "X::" prefixes in random places in hope to work around future problems).

> if this does go wrong [a disaster beyond your imagination will occur]

If omitting `Ssl::` from `Ssl::MaybeSetupRsaCallback()` compiles but causes runtime bugs, then we need to fix that (and/or some other) API (instead of hoping that some human will predict and overt that disaster by adding "Ssl::" prefixes in all the right places).